### ISSUE TITLE
[core] Conflicting contracts of getChild between PMD & Antlr

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#2235]
The contract of `Node::getChild` (#2172) is to throw when the index is out of bounds. Antlr parse trees also happen to have a `getChild` method whose contract is to return null in this case. Should we align ourselves? I think it's reasonable, though this needs to be pointed out in the release notes, as it's a noticeable departure from the current status quo

**Comment by adangel** [comment]
I think, the underlying issue here is, that the AntlrBaseNode (#1658) extends a antlr class (`ParserRuleContext`) and the AntlrNode interface extends `ParseTree` from antlr. Since we decided, that the AST of a language that PMD supports is one of *our* API, we shouldn't do this. We should rather use composition then. Also, I'm not sure, whether we need to expose the fact in the AST, that the implementation is Antlr - so AntlrNode shouldn't be public API at all. I consider this an implementation detail.

Putting that aside (it's a different issue): What would make most sense for `Node::getChild`? In the impl, we in any case need to check the bounds. Throwing a exception might induce using try-catch statements just to avoid an extra check:

```java
// variant 1 - we should probably not foster this style
Node root = ...;
Node child = null;
try {
  child = root.getChild(42);
} catch (IndexOutOfBoundsException ignore) {
  // ignored
}
if (child != null) {
    System.out.println("found it!");
}
```

vs.

```java
// variant 2 - more verbose, avoiding the exception
Node root = ...;
Node child = null;
if (root.getNumChildren() > 42) { // error prone: is it > or >=? 0-based index, 1-based index?
  child = root.getChild(42);
}
if (child != null) {
    System.out.println("found it!");
}
```

vs.

```java
// variant 3 - shorter
Node root = ...;
Node child = root.getChild(42);
if (child != null) {
    System.out.println("found it!");
}
```

So, in summary, I agree that it's more reasonable to return `null` in that case.

**Comment by oowekyala** [comment]
I've kind of reconsidered my stance of this. I think, your example 2 is how the API should be used. There's nothing really confusing about it, it works just like the collections API or arrays:
```java
if (list.size() >= 2) return list.get(0) + list.get(1);
```
Besides, asking for a child at a random index like 42 is pretty rare in our sources, and *should* be done carefully. The most used use-cases for getChild are:
1. iterating all children
1. getting the first child
1. getting the last child

This sums it up pretty well:

![Capture d’écran de 2020-04-27 21-08-40](https://user-images.githubusercontent.com/24524930/80410725-5a1fe600-88cb-11ea-956e-052fa0e43c7f.png)

Use case (1)  is covered by the `children()` node stream. The two others have a method on java-grammar, which I pulled up into #2447 (getLastChild, getFirstChild, both of which are nullable because yes, it's a bit less verbose). 

But returning null for *any* out-of-bounds may hide programming errors (like confusing `>=` for `>` in your example 2). And what if the index is negative? That would definitely be a programming error. I'd rather have an IndexOutOfBounds immediately than an NPE later on, which wouldn't mention the indices (and an NPE could be any number of things).

If someone needs a `getChild` with nullable semantics, they can always do `children().get(42)`: NodeStream would return null if it's out-of-bounds.

I agree though that the real problem is how the Antlr support was implemented. There are many other problems with it, that cannot be solved if our model is just to extend antlr classes.

**Comment by adangel** [comment]
So, I think (please correct me if I'm wrong), we settled for the following:

The API `Node::getChild(int)` stays as is:

```java
    /**
     * Returns the child of this node at the given index.
     *
     * @throws IndexOutOfBoundsException if the index is negative or greater than {@link #getNumChildren()}.
     */
    Node getChild(int index);
```

Which means, it *will* throw a IndexOutOfBoundsException if an invalid index is used.

Therefore I'm closing this issue. The other part concerning ANTLR is relevant for PMD 7 only.
